documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
-+       /// </summary>
-+       /// <param name="handle">The native handle created/sourced from the underlying native API used by this input backend.</param>
-+       /// <remarks>
-+       /// This method is implicitly called by the Windowing-Input integration. <br />
-+       /// On desktop, this uses GLFW. The handle refers to a <c>GLFWwindow*</c> if GLFW is in use.
-+       /// </remarks>
-+       public static unsafe IInputBackend Create(void* handle);
-+   }
+- This proposal incorporates/supersedes the Enhanced Input Events proposal.
+- This proposal assumes knowledge of the Windowing 3.0 proposal.
+- This proposal assumes knowledge of Silk.NET.Core 2.0.
+
+Please note that text marked **INFORMATIVE TEXT** does not form part of this proposal's requirements or proposed API, and is merely text to help the reader understand our intentions. Informative text is not well-defined, and may not reflect the actual implementation.
+
+Note that the parties in this proposal are:
+- **developer**: The developer using Silk.NET.
+- **end user**: The user using an application using Silk.NET, and by extension the user inputting data into the application via Silk.NET.
+
+Cases where the **user** word is used without the **end** prefix can be assumed to be referring to the **developer** rather than the **end user**.
+
+# Usage Examples
+```cs
+IWindowHandlesSource someWindow = null!;
+var inputContext = someWindow.CreateInput();
+inputContext.Gamepads.ThumbstickMove += @event =>
+{
+    Console.WriteLine($"Thumbstick {@event.Index} moved from {@event.OldValue} to {@event.NewValue}");
+};
+var isButtonDown = inputContext.Gamepads.Any(gamepadState => gamepadState.Buttons[JoystickButton.A]);
+```
+```cs
+IWindowHandlesSource someWindow = null!;
+var inputContext = new InputContext();
+inputContext.Backends.Add(someWindow.CreateInputBackend());
+// in future:
+// inputContext.Backends.Add(new OpenXRInputBackend(...));
+```
+
+# Reference Implementation
+
+Similar to Windowing 3.0, a reference implementation will be included in the main `Silk.NET.Input` package which uses the same API or family of APIs as Windowing's reference implementation. This will be exposed by the `InputWindowExtensions` class.
+
+```cs
+public static class InputWindowExtensions
+{
+    public static IInputBackend CreateInputBackend(this WindowHandles window);
+    public static IInputBackend CreateInputBackend(this IWindowHandlesSource window);
+    public static InputContext CreateInput(this WindowHandles window);
+    public static InputContext CreateInput(this IWindowHandlesSource window);
+}
+```
+
+The `CreateInputBackend` will create an instance of the reference implementation for the given `WindowHandles`. The `IWindowHandlesSource` overloads just forward to the `WindowHandles` overload. This is because `ISurface` will implement `IWindowHandlesSource`, so the extension methods will be usable on an `ISurface` without having a hard reference between Windowing and Input.
```

Comment:
(I can only mark the whole line)
I don't understand the reference to ISurface here, why is this relevant to this proposal?
imo IWindowHandlesSource / IWindowHandles should be re-defined in this proposal and no reference to ISurface is needed here.

Reply:
The Windowing proposal also uses these types - in any case we have a circular dependency, and of the Input and Windowing proposals, window-related types feel more at home in the latter.